rfcs/0004-checkpoints.md

```diff
@@ -0,0 +1,315 @@
+5. Set `writer_epoch` to E
+6. Write the manifest. If this fails with a CAS error, go back to 1.
+
+#### Refreshing the Checkpoint
+
+The writer will periodically re-establish its checkpoint to pin the latest SSTs and SRs. It will run the following whenever it detects that the manifest has been updated. This happens either when the manifest is polled (at `manifest_poll_interval`), or when the writer detects an update when it goes to update the manifest with it’s local view of the core db state:
+1. Read the latest manifest version V. (If the writer is applying an update from the local state then in the first iteration it can assume that it’s local version is the latest. If it’s wrong, the CAS will fail and it should reload the manifest from Object Storage))
+2. Decide whether it needs to re-establish it’s checkpoint. The writer reestablishes the checkpoint whenever the set of L0 SSTs, or Sorted Runs has changed. This includes both changes made locally by the Writer, or by the Compactor.
+    1. If so, create a new checkpoint with `manifest_id` set to V+1, `checkpoint_expire_time` set to None, and `metadata` set to `WriterCheckpoint{ epoch: E }`.
+    2. Delete the old checkpoint from `checkpoints`. When we support range scans, we’ll instead want to maintain references to the checkpoint from the range scans and delete the checkpoint when all references are dropped. (We should technically do this for point lookups too, but the likelihood of GC deleting an SST in this case is very low).
+    3. Apply any local changes
+    4. Write the new manifest. If this fails with a CAS error, go back to 1.
+
+#### Clones
+
+The writer will also support initializing a new database from an existing checkpoint. This allows users to “fork” an instance of slatedb, allowing it to access all the original db’s data from the checkpoint, but isolate writes to a new db instance.
+
+To support this, we add the `Db::open_from_checkpoint` method, which accepts a `source_path` and `source_checkpoint`. When initializing the db for the first time (detected by the presence of a manifest), the writer will do the following:
+1. Read the source manifest MC at `source_checkpoint`.
+2. Copy over any WAL SSTs between `last_compacted_wal_id` and `wal_id_last_seen`
+3. Verify that `source_checkpoint` is present and not expired. Create a new checkpoint in the source database with the `manifest_id` as `source_checkpoint`.
+4. Compute an initial manifest for the new DB. The new manifest will set its fields as follows
+    - writer_epoch: set to the epoch from the source checkpoint + 1.
+    - compactor_epoch: copied from the source checkpoint.
+    - wal_id_last_compacted: copied from the source checkpoint
+    - wal_id_last_seen: copied from the source checkpoint
+    - l0: copied from the source checkpoint
```

Comment:
Why do we need to copy `l0`? Isn't it possible to use it from the source manifest?

I see SST Path resolution in next section. Isn't it needed only for the first manifest? With that, if we don't copy l0, compacted over to new manifest, wouldn't it be possible to not do SST path resolution?

Reply:
> Why do we need to copy l0? Isn't it possible to use it from the source manifest?
I don't mean that the actual L0 files are copied. I'm just saying that the contents of the l0 field are copied into the new manifest.

> I see SST Path resolution in next section. Isn't it needed only for the first manifest? With that, if we don't copy l0, compacted over to new manifest, wouldn't it be possible to not do SST path resolution?

The SST files themselves are not copied and remain at their original path in the object store. The clone would need to be able to find SST files from the parent db when doing reads or compacting.

Reply:
I understand the part about not copying the actual SSTs. I am asking if the `l0` with the list of ssts from original DB need to be copied over, instead of using the original db's `l0` directly.

Reply:
Oh, you mean reading it out of the source db's manifest? I suppose we could but it would make it more complicated to resolve. For example, if you append a new l0 in the clone manifest, you'd need to know to merge the clone l0 list with the list from the source manifest. Basically, it seems much more straightforward to resolve the path of a unique and immutable SST object over multiple source locations than it is to reconcile the constantly changing `l0` and `compacted` lists.

Reply:
Yes, i mean reading it out of source's manifest. I agree resolving path is simpler.